tortoisebeacon/tortoise_beacon.go

```diff
@@ -592,59 +548,143 @@ func (tb *TortoiseBeacon) runConsensusPhase(ctx context.Context, epoch types.Epo
 	ticker := time.NewTicker(tb.config.VotingRoundDuration + tb.config.WeakCoinRoundDuration)
 	defer ticker.Stop()
 
-	var coinflip bool
-	for round := firstRound; round <= tb.config.RoundsNumber; round++ {
+	var (
+		coinFlip            uint64
+		ownLastRoundVotesMu sync.RWMutex
+		ownLastRoundVotes   allVotes
+	)
+
+	for round := firstRound; round <= tb.lastRound(); round++ {
 		// always use coinflip from the previous round for current round.
 		// round 1 is running without coinflip (e.g. value is false) intentionally
 		round := round
-		tb.tg.Go(func(ctx context.Context) error {
-			if err := tb.sendVotes(ctx, epoch, round, coinflip); err != nil {
-				tb.Log.With().Error("Failed to send voting messages",
+		err := tb.tg.Go(func(ctx context.Context) error {
+			if round == firstRound {
+				if err := tb.sendProposalVote(ctx, epoch); err != nil {
+					tb.Log.With().Error("Failed to send proposal vote",
+						log.Uint32("epoch_id", uint32(epoch)),
+						log.Uint32("round_id", uint32(round)),
+						log.Err(err))
+
+					return fmt.Errorf("failed to send proposal vote: %w", err)
+				}
+
+				return nil
+			}
+
+			// next rounds, send vote
+			// construct a message that points to all messages from previous round received by δ
+			ownCurrentRoundVotes, err := tb.calcVotes(epoch, round, atomic.LoadUint64(&coinFlip) != 0)
+			if err != nil {
+				tb.Log.With().Error("Failed to calculate votes",
 					log.Uint32("epoch_id", uint32(epoch)),
-					log.Uint64("round_id", uint64(round)),
+					log.Uint32("round_id", uint32(round)),
 					log.Err(err))
+
+				return fmt.Errorf("calculate votes: %w", err)
+			}
+
+			if round == tb.lastRound() {
+				ownLastRoundVotesMu.Lock()
+				ownLastRoundVotes = ownCurrentRoundVotes
+				ownLastRoundVotesMu.Unlock()
 			}
+
+			if err := tb.sendFollowingVote(ctx, epoch, round, ownCurrentRoundVotes); err != nil {
+				tb.Log.With().Error("Failed to send following vote",
+					log.Uint32("epoch_id", uint32(epoch)),
+					log.Uint32("round_id", uint32(round)),
+					log.Err(err))
+
+				return fmt.Errorf("send following vote: %w", err)
+			}
+
 			return nil
 		})
-		tb.tg.Go(func(ctx context.Context) error {
-			tb.startWeakCoin(ctx, epoch, round)
+		if err != nil {
+			tb.Log.Warning("taskgroup: Go returned an error",
+				log.Err(err))
+		}
+
+		err = tb.tg.Go(func(ctx context.Context) error {
+			tb.startWeakCoinRound(ctx, epoch, round)
 			return nil
 		})
+		if err != nil {
+			tb.Log.Warning("taskgroup: Go returned an error",
+				log.Err(err))
+		}
+
 		select {
 		case <-ticker.C:
 		case <-ctx.Done():
-			return false
+			return allVotes{}
 		}
+
 		tb.weakCoin.FinishRound()
-		coinflip = tb.weakCoin.Get(epoch, round)
+
+		cf := tb.weakCoin.Get(epoch, round+1)
+		if cf {
+			atomic.StoreUint64(&coinFlip, 1)
```

Comment:
if atomic is required this is clearly a bug. coinflip from the previous round should be used in current round, concurrent update doesn't make any sense

Reply:
Without it, a data race is reported in `TestTortoiseBeacon`

Reply:
then we either need to wait until goroutine from previous rounds exited, or copy coinflip before it is passed to a goroutine

```go
t.start(coinflip bool) {
  tb.tg.Go(func() {
   // use coinflip
  })
}
```

this way you will use correct version of the coinflip

Reply:
the race here means that you can use coinFlip in round 2 that is meant to be used in round 3. so solving the race condition will only hide the problem but not solve it

Reply:
@dshulyak Could you please look if https://github.com/spacemeshos/go-spacemesh/pull/2649/commits/48914176493d3a1e488df0996fdb83c55154e4a3 contains what you meant?

Reply:
this should work too :+1: . i think that it would be cleaner with a function, but whatever you prefer